cmd/server/artifacts/artifact_server.go

```diff
@@ -0,0 +1,183 @@
+	workflowName := path[3]
+	nodeId := path[4]
+	artifactName := path[5]
+
+	log.WithFields(log.Fields{"namespace": namespace, "workflowName": workflowName, "nodeId": nodeId, "artifactName": artifactName}).Info("Download artifact")
+
+	wf, err := a.getWorkflow(ctx, namespace, workflowName)
+	if err != nil {
+		a.serverInternalError(err, w)
+		return
+	}
+
+	data, err := a.getArtifact(ctx, wf, nodeId, artifactName)
+	if err != nil {
+		a.serverInternalError(err, w)
+		return
+	}
+	a.ok(w, data)
+}
+func (a *ArtifactServer) GetArtifactByUID(w http.ResponseWriter, r *http.Request) {
+
+	ctx, err := a.gateKeeping(r)
+	if err != nil {
+		w.WriteHeader(401)
+		_, _ = w.Write([]byte(err.Error()))
+		return
+	}
+	path := strings.SplitN(r.URL.Path, "/", 6)
+
+	uid := path[2]
+	nodeId := path[3]
+	artifactName := path[4]
+
+	log.WithFields(log.Fields{"uid": uid, "nodeId": nodeId, "artifactName": artifactName}).Info("Download artifact")
+
+	wf, err := a.getWorkflowByUID(ctx, uid)
+	if err != nil {
+		a.serverInternalError(err, w)
+		return
+	}
+
+	data, err := a.getArtifact(ctx, wf, nodeId, artifactName)
+	if err != nil {
+		a.serverInternalError(err, w)
+		return
+	}
+	a.ok(w, data)
+}
+func (a *ArtifactServer) gateKeeping(r *http.Request) (context.Context, error) {
+	// TODO - we should not put the token in the URL - OSWAP obvs
+	authHeader := r.URL.Query().Get("Authorization")
+	ctx := metadata.NewIncomingContext(r.Context(), metadata.MD{"grpcgateway-authorization": []string{authHeader}})
+	return a.authN.Context(ctx)
+}
+
+func (a *ArtifactServer) ok(w http.ResponseWriter, data []byte) {
+	w.WriteHeader(200)
+	_, err := w.Write(data)
+	if err != nil {
+		w.WriteHeader(500)
+		_, _ = w.Write([]byte(err.Error()))
+		return
+	}
+}
+
+func (a *ArtifactServer) serverInternalError(err error, w http.ResponseWriter) {
+	w.WriteHeader(500)
+	_, _ = w.Write([]byte(err.Error()))
+}
+
+func (a *ArtifactServer) getArtifact(ctx context.Context, wf *wfv1.Workflow, nodeId, artifactName string) ([]byte, error) {
+	kubeClient := auth.GetKubeClient(ctx)
+
+	art := wf.Status.Nodes[nodeId].Outputs.GetArtifactByName(artifactName)
+	if art == nil {
+		return nil, fmt.Errorf("artifact not found")
+	}
+
+	driver, err := artifact.NewDriver(art, resources{kubeClient, wf.Namespace})
+	if err != nil {
+		return nil, err
+	}
+
+	tmp, err := ioutil.TempFile(".", "artifact")
+	if err != nil {
+		return nil, err
+	}
+	path := tmp.Name()
+	defer func() { _ = os.Remove(path) }()
+
+	err = driver.Load(art, path)
+	if err != nil {
+		return nil, err
+	}
+
+	file, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	log.WithFields(log.Fields{"size": len(file)}).Debug("Artifact file size")
+
+	return file, nil
+}
+
+func (a *ArtifactServer) getWorkflow(ctx context.Context, namespace string, workflowName string) (*wfv1.Workflow, error) {
+	wfClient := auth.GetWfClient(ctx)
+	wf, err := wfClient.ArgoprojV1alpha1().Workflows(namespace).Get(workflowName, metav1.GetOptions{})
```

Comment:
auth-z is not being performed here, nor is it done by the callers. this means artifacts are able to be downloaded by anyone -- a security hole.

Reply:
I don't think you are right on its - line 152 mean you must be able to read the workflow.

Reply:
You're right, disregard.